File: src/main/java/io/bonitoo/client_vs_http/AbstractIOTWriter.java
package io.bonitoo.client_vs_http;

import java.io.InterruptedIOException;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import io.bonitoo.influxdb.reactive.InfluxDBReactiveFactory;
import io.bonitoo.influxdb.reactive.options.InfluxDBOptions;

import org.apache.commons.cli.CommandLine;
import org.influxdb.dto.Query;

/**
 * @author Jakub Bednar (26/09/2019 12:21)
 */
@SuppressWarnings("CatchMayIgnoreException")
abstract class AbstractIOTWriter {

    private final String measurementName;
    private final int threadsCount;
    private final int secondsCount;
    private final int lineProtocolsCount;
    private final int expectedCount;
    private volatile boolean execute = true;

    AbstractIOTWriter(CommandLine line) {

        measurementName = "sensor_" + System.currentTimeMillis();
        threadsCount = Integer.parseInt(line.getOptionValue("threadsCount", "2000"));
        secondsCount = Integer.parseInt(line.getOptionValue("secondsCount", "30"));
        lineProtocolsCount = Integer.parseInt(line.getOptionValue("lineProtocolsCount", "100"));
        expectedCount = threadsCount * secondsCount * lineProtocolsCount;

        System.out.println("measurement:        " + measurementName);
        System.out.println("threadsCount:       " + threadsCount);
        System.out.println("secondsCount:       " + secondsCount);
        System.out.println("lineProtocolsCount: " + lineProtocolsCount);
        System.out.println();
    }

    AbstractIOTWriter start() {

        System.out.println("expected size: " + expectedCount);
        System.out.println();

        ExecutorService executor = Executors.newFixedThreadPool(threadsCount);

        for (int i = 1; i < threadsCount + 1; i++) {
            Runnable worker = new IOTSensor(i, measurementName, secondsCount, lineProtocolsCount);
            executor.execute(worker);
        }
        executor.shutdown();

        long start = System.currentTimeMillis();

        // Wait until all threads are finish
        while (!executor.isTerminated()) {
            try {
                Thread.sleep(100);
            } catch (InterruptedException e) {
            }

            //
            // Stop benchmark after elapsed time
            //
            if (System.currentTimeMillis() - start > secondsCount * 1_000 && execute) {
                System.out.println("\n\nThe time: " + secondsCount + " seconds elapsed! Stopping all writers");
                execute = false;
                executor.shutdownNow();
            }
        }

        System.out.println();
        System.out.println();

        finished();

        return this;
    }

    void verify() {
        InfluxDBOptions options = InfluxDBOptions.builder()
                .url("http://localhost:8086")
                .database("iot_writes")
                .build();

        InfluxDBReactiveFactory.connect(options)
                .query(new Query("select count(*) from " + measurementName, "iot_writes"))
                .subscribe(queryResult -> {

                    Double count = (Double) queryResult.getResults().get(0).getSeries().get(0).getValues().get(0).get(1);

                    System.out.println("Results:");
                    System.out.println("-> expected:    " + expectedCount);
                    System.out.println("-> total:       " + count);
                    System.out.println("-> rate:        " + (count / expectedCount) * 100);
                });
    }

    abstract void writeRecord(final String records) throws Exception;

    abstract void finished();

    private class IOTSensor implements Runnable {

        private final Integer id;
        private final String measurementName;
        private final int secondCount;
        private final int lineProtocolsCount;

        IOTSensor(Integer id, String measurementName, final int secondCount, final int lineProtocolsCount) {
            this.id = id;
            this.measurementName = measurementName;
            this.secondCount = secondCount;
            this.lineProtocolsCount = lineProtocolsCount;
        }

        @Override
        public void run() {

            try {
                doLoad();
            } catch (Exception e) {
                if (e instanceof InterruptedException || e instanceof InterruptedIOException) {
                    return;
                }
                System.out.println("e.getMessage() = " + e.getMessage());
            }
        }

        private void doLoad() throws Exception {
            for (int i = 0; i < secondCount && execute; i++) {

                if (!execute) {
                    break;
                }

                //
                // Logging
                //
                if (id == 1) {
                    System.out.print(String.format("\rwriting iterations: %s/%s ", i + 1, secondCount));
                }

                //
                // Generate data
                //
                int start = i * lineProtocolsCount;
                int end = start + lineProtocolsCount;

                List<String> records = IntStream
                        .range(start, end)
                        .mapToObj(j -> String.format("%s,id=%s temperature=%d %d", measurementName, id, System.currentTimeMillis(), j))
                        .collect(Collectors.toList());

                //
                // Write records one by one
                //
                for (String record : records) {
                    if (execute) {
                        writeRecord(record);
                    }
                }

                if (!execute) {
                    break;
                }

                Thread.sleep(1000);
            }
        }
    }
}

File: csharp/Benchmark/AbstractIotWriter.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Diagnostics;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.CommandLineUtils;

namespace Benchmark
{
    public abstract class AbstractIotWriter
    {
        protected readonly string MeasurementName;
        protected readonly int ThreadsCount;
        protected readonly int SecondsCount;
        protected readonly int LineProtocolsCount;
        protected readonly int ExpectedCount;
        protected static bool Execute = true;
        protected readonly bool SkipCount;
        public static int Counter = 0;

        public AbstractIotWriter(List<CommandOption> options)
        {
            MeasurementName = Benchmark.GetOptionValue(GetOption(options, "measurementName"),
                "sensor_" + CurrentTimeMillis());
            ThreadsCount = int.Parse(Benchmark.GetOptionValue(GetOption(options, "threadsCount"), "2000"));
            SecondsCount = int.Parse(Benchmark.GetOptionValue(GetOption(options, "secondsCount"), "60"));
            LineProtocolsCount = int.Parse(Benchmark.GetOptionValue(GetOption(options, "lineProtocolsCount"), "100"));
            SkipCount = GetOption(options, "skipCount").HasValue();
            ExpectedCount = ThreadsCount * SecondsCount * LineProtocolsCount;

            Console.WriteLine("measurement:        " + MeasurementName);
            Console.WriteLine("threadsCount:       " + ThreadsCount);
            Console.WriteLine("secondsCount:       " + SecondsCount);
            Console.WriteLine("lineProtocolsCount: " + LineProtocolsCount);
            Console.WriteLine();
        }

        public async Task<AbstractIotWriter> Start()
        {
            Console.WriteLine("expected size: " + ExpectedCount);
            Console.WriteLine();

            var cancellationTokenSource = new CancellationTokenSource();
            cancellationTokenSource.CancelAfter(SecondsCount);

            var threads = new Collection<Thread>();
            for (int i = 0; i < ThreadsCount; i++)
            {
                Thread t = new Thread(new ParameterizedThreadStart(DoLoad));
                threads.Add(t);
                t.Start(i);
            }

            for (int i = 0; i < threads.Count; i++)
            {
                var thread = threads[i];
                thread.Join(SecondsCount * 1000);
            }

            Console.WriteLine("Writer counter: {0}", Counter);
            
            Console.WriteLine();
            Console.WriteLine();

            Finished();

            return this;
        }

        public void Verify()
        {
            if (SkipCount)
            {
                return;
            }

            var count = CountInDb().GetAwaiter().GetResult();

            Console.WriteLine("Results:");
            Console.WriteLine("-> expected:        " + ExpectedCount);
            Console.WriteLine("-> total:           " + count);
            Console.WriteLine("-> rate [%]:        " + (count / ExpectedCount) * 100);

            Console.ForegroundColor = ConsoleColor.Green;
            Console.WriteLine("-> rate [msg/sec]:  " + count / SecondsCount);
            Console.ResetColor();
        }

        protected abstract Task<double> CountInDb();

        protected abstract void WriteRecord(string records);

        protected abstract void Finished();

        public CommandOption GetOption(List<CommandOption> options, string type)
        {
            return options.FindAll(o => type.Equals(o.ShortName))[0];
        }

        public static long CurrentTimeMillis()
        {
            var jan1St1970 = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);

            return (long) (DateTime.UtcNow - jan1St1970).TotalMilliseconds;
        }

        private void DoLoad(object param)
        {
            int id = Convert.ToInt32(param);
            Console.Write("S:{0}",id);
            Stopwatch stopwatch = new Stopwatch();
            stopwatch.Start();
            var random = new Random();
            for (var ii = 0; ii < SecondsCount && Execute; ii++)
            {
                if (stopwatch.ElapsedMilliseconds >= SecondsCount * 1000)
                {
//                    Console.WriteLine("Time elapsed for thread: {0}, id={1}", Thread.CurrentThread.ManagedThreadId, id);
                    break;
                }

                if (!Execute)
                {
                    break;
                }

                //
                // Logging
                //
                if (id == 1)
                {
                    Console.WriteLine("writing iterations: " + (ii + 1) + "/" + SecondsCount);
                }
                
//                Console.Write(".");

                //
                // Generate data
                //
                var start = ii * LineProtocolsCount;
                var end = start + LineProtocolsCount;

                var records = new List<string>();
                for (int j = start; j < end; j++)
                {
                    var record = MeasurementName + "," + "id=" + id + " temperature="
                                 + random.Next(0, Int32.MaxValue) + " " + j;
                    records.Add(record);
                }

                //
                // Write records one by one
                //
                foreach (var record in records)
                {
                    if (Execute)
                    {
                        WriteRecord(record);
                    }
                }

                if (!Execute)
                {
                    break;
                }

                Thread.Sleep(1000);
            }
            Console.Write("E:{0} ", id);
        }
    }
}